### DISCUSSION TITLE
测试讨论区

### DISCUSSION THREAD

**Discussion by afc163** [ant-design/ant-design#36882] [General]
测试讨论区

**Comment by clean99** [comment]
Hi, 我最近在学习测试，看了一下antd的codebase，发现一些问题想与你们讨论一下：
1. 使用`fireEvent`而不是`userEvent`：`react-testing-library`的作者更加推荐`userEvent`因为它能模仿用户与组件交互的过程，比如`fireEvent.change`只会触发改变`input`的事件，而`userEvent.type`会触发`keyDown`, `keyPress`, and `keyUp`事件。
2. 使用`querySelector`获取元素(不使用screen)：`react-testing-library`的作者推荐使用`screen.getByRole`...来做查询，因为如果使用querySelector会让测试可读性变差，而且这样做避开了accessibility的检查，比如getByRole会同时保证role的正确性。[About Query](https://testing-library.com/docs/queries/about/#priority)
3. act()报错：我运行`npm run test`的时候，会有一堆`console.error`
>  
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

这个应该是有些`jest.runAllTimers`没有包在act里面? 还有一些地方可能用了`setState`更新状态，但是query的时候没有用`waitFor`也会有这个报错。
4. 大量的使用`container`与`snapshot`:[button test](https://github.com/ant-design/ant-design/blob/master/components/button/__tests__/index.test.tsx#L65)比如说检查中文字符是否渲染在页面上，推荐用`screen.getByRole('button, {.name: /按钮/i }')`而不是使用`container.firstChild`。另外我认为使用`snapshot`严重影响测试的可读性，如果使用推荐使用`toMatchInlineSnapshot`，这样读者至少可以直接看到dom。
5. 有些测试似乎测试了不必要的东西：[比如这个](https://github.com/ant-design/ant-design/blob/c0b9b07f87e73388653df0ccfcc0f1668ba2bce0/components/date-picker/__tests__/RangePicker.test.js#L46) 这个测试实际上只测试了Test组件的实现，相当于创建了一个`Test User`。测试应该总是测试`Product User` 和 `Develop User`， 这个测试应该`render`一个`RangePicker`，然后测试`onPanelChange`的时候如何调用函数比较合适？

想讨论一下以上这些问题你们觉得有必要改吗？如果需要的话我可以帮忙重构一下部分测试，谢谢～

**Reply by clean99** [reply]
另外，还有部分使用`Enzyme`测试，这个库好像已经不维护了，而且这个库提供的抽象让开发者很容易就去测试实现，如果代码重构修改，测试很可能会失败，另外也没办法保证user看到的东西被测试过。

**Reply by afc163** [reply]
@clean99 直接来 PR 吧，一个问题一个 PR 分开

**Reply by clean99** [reply]
> @clean99 直接来 PR 吧，一个问题一个 PR 分开

好的，感谢回复，我后面提点pr～
